### PR TITLE
Add AP_Filesystem API

### PR DESCRIPTION
**Pull request by tridge**
This adds an AP_FIlesystem API for all filesystem operations. It cleans up the existing posix abstraction in ChibiOS, and will enable us to add automatic remount on operation failures
A new parameter LOG_FILE_TIMEOUT is added which gives the timeout in seconds for log writes. Setting this to 30 allows you to remove the card for 30s and when re-inserted logging will continue.
This is an alternative to #8639 from @peterbarker 

Issues:

 - [x] adds about 1.7k to flash for unknown reason (fixed, is now smaller than master)
 - [x] add automatic remount on IO error

### CODE REVIEW

**Comment by tridge** [comment]
@WickedShell please check if the AP_Scripting changes are OK (also need to check it all works).
Note that this fixes the recursive opendir issue

File: libraries/AP_Terrain/TerrainIO.cpp [hunk 1]
```diff
@@ -196,13 +189,9 @@ void AP_Terrain::open_file(void)
...
     if (fd != -1) {
-        ::close(fd);
+        AP::FS().close(fd);
     }
-#if HAL_OS_POSIX_IO
-    fd = ::open(file_path, O_RDWR|O_CREAT|O_CLOEXEC, 0644);
-#else
-    fd = ::open(file_path, O_RDWR|O_CREAT|O_CLOEXEC);
-#endif
+    fd = AP::FS().open(file_path, O_RDWR|O_CREAT|O_CLOEXEC);
```
Comment (sh83) [inline libraries/AP_Terrain/TerrainIO.cpp]: I wonder why 'O_CLOEXEC' is here?
It is not handled and I think this flag is used for sockets and not for posix open.

Comment (tridge) [reply libraries/AP_Terrain/TerrainIO.cpp]: fixed (and it is relevant for files, but not on ChibiOS)

**Comment by tridge** [comment]
this was flight tested today on a bixler

**Comment by tridge** [comment]
@WickedShell ok, thanks. I'll fix this soon

**Comment by tridge** [comment]
@WickedShell I've now got it working with Lua scripts, and also now have a common FILE implementation between SITL and ChibiOS, so valgrind does useful checking of Lua IO calls

File: libraries/AP_Scripting/lua_scripts.cpp [hunk 2]
```diff
@@ -157,7 +149,7 @@ void lua_scripts::load_all_scripts_in_dir(lua_State *L, const char *dirname) {
         reschedule_script(script);
 
     }
-    closedir(d);
+    AP::FS().closedir(d);
```
Comment (WickedShell) [inline libraries/AP_Scripting/lua_scripts.cpp]: We may never hit this, because load_script may never return here if there is an error within the script. This would leak the dir. I think we can resolve this by making the pointer be a part of the `lua_scripts` class, then in the panic handler we can check if it's not `nullptr` and delete it then. (As long as we always set it to be `nullptr` after deleting it). This is a small leak at the moment, but in the future we are considering reloading scripts more frequently.

Comment (tridge) [reply libraries/AP_Scripting/lua_scripts.cpp]: or we could add a AP::FS() method that does does a opendir with a pre-allocated piece of memory or with a realloc() function as a univeral allocator fn

Comment (tridge) [reply libraries/AP_Scripting/lua_scripts.cpp]: @WickedShell isn't this an existing bug anyway? (at least on HAL_Linux, where opendir does alloc)

File: libraries/AP_Proximity/AP_Proximity_LightWareSF40C.cpp [hunk 3]
```diff
@@ -383,8 +383,8 @@ bool AP_Proximity_LightWareSF40C::process_reply()
         case RequestType_Health:
             // expect result in the form "0xhhhh"
             if (element_len[0] > 0) {
-                int result;
-                if (sscanf(element_buf[0], "%x", &result) > 0) {
+                long int result = strtol(element_buf[0], nullptr, 16);
```
Comment (davidbuzz) [inline libraries/AP_Proximity/AP_Proximity_LightWareSF40C.cpp]: I'm not sure why this change to the proximity sensor is included in the AP_Filesystem PR.?   I haven't looked into the implications of the change from sscanf to strtol, and it's probably a harmless change, but thought I'd note it here anyway.

**Comment by davidbuzz** [comment]
I've read through all the code changes in the AP_Filesystem PR, and had just one comment that I added above. everything else looks like it's a pretty straight-forward relocation of the existing filesystem functions into the AP_FS:: namespace without too much to worry about.
 ... although it does touch a lot of files.

**Comment by tridge** [comment]
merged after discussion with mdb
